packages/test/test-utils/src/opProcessingController.ts

```diff
@@ -101,26 +137,34 @@ class DeltaManagerMonitor extends DeltaManagerToggle {
             // Clear the pending count and start anew
             this.pendingCount = 0;
             this.firstClientSequenceNumber = -1;
+            this.lastOutbound = undefined;
         });
         deltaManager.outbound.on("op", this.outbound.bind(this));
         deltaManager.inbound.on("push", this.inbound.bind(this));
     }
 
     public get latestSequenceNumber() {
-        return this._latestSequenceNumber;
+        return this.deltaManager.lastSequenceNumber;
     }
 
     public hasPendingWork() {
-        // If we are not connected, we assume that we are trying to, and classify as pending work
-        return this.clientId === undefined || this.pendingCount !== 0;
+        return !this.deltaManager.disposed
+            && (this.pendingWriteConnection || this.pendingCount !== 0);
     }
 
     private connect(clientId: string) {
         this.clientId = clientId;
         this.trace("CON");
     }
     private inbound(message: ISequencedDocumentMessage) {
-        this._latestSequenceNumber = message.sequenceNumber;
+        if (message.clientId) {
+            this.lastInboundPerClient.set(message.clientId, message);
+        }
+        if (message.type === MessageType.ClientLeave) {
+            const systemLeaveMessage = message as ISequencedDocumentSystemMessage;
+            const clientId = JSON.parse(systemLeaveMessage.data) as string;
+            this.lastInboundPerClient.delete(clientId);
+        }
 
         if (this.clientId === undefined) {
             // Ignore message when we are not connected.
```

Comment:
I do not know if it matters, but logic below comparing of message.clientId !== this.clientId does not correctly account how disconnect/reconnect work, because we register for wrong event in constructor:

deltaManager.on("connect", (details) => this.connect(details.clientId));

clientId should be switched only when Container raises "connected" event, which happens only after inbouding all ops  up to own join op. Delta Manager's "connect" fires too early and is not propagated to IRuntime, it should only be used by hosts (exposed from Container as "connect" event) for connected/disconnected banner logic.

I've opened an issue somewhere (assigned to Matt) to reevaluate slit of logic between Container & DeltaManager, as it's should be in one place (today, while DeltaManager sort of does not know about this, it actually does - there are couple places where it does unnatural things to account for this discrepancy).

Reply:
yeah. the op processing controller is in a bit of a weird place. in general, i'd like to move away from it, as it's not reliable for a lot of the reasons you call out. in general i'd rather see tests await more deterministic side effects but refactoring for that is not something i think we have time for currently.